commit 88369bd27cf8c5fd8d83574157a097aa557d499b
Author: Tatsuya Fukushi <fukushi.tatsuya@fujitsu.com>
Date:   Tue Sep 8 08:19:17 2020 +0000

    do not using ctransrate

diff --git a/setup.py b/setup.py
index 230d8bc2..4fa3767a 100644
--- a/setup.py
+++ b/setup.py
@@ -43,13 +43,13 @@ setup(
     keywords="tensorflow opennmt nmt neural machine translation",
     python_requires='>=3.5',
     install_requires=[
-        "ctranslate2>=1.7,<2;platform_system=='Linux'",
+        #"ctranslate2>=1.7,<2;platform_system=='Linux'",
         "pyonmttok>=1.18.1,<2;platform_system=='Linux'",
         "pyyaml>=5.3,<5.4",
         "rouge>=1.0,<2",
         "sacrebleu>=1.4.9,<2",
-        "tensorflow>=2.2,<2.3",
-        "tensorflow-addons>=0.10,<0.11",
+        "tensorflow>=2.2,<=2.7",
+        "tensorflow-addons>=0.10,<=0.15",
         "pyter3==0.3"
     ],
     extras_require={
